keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,398 @@
+
+With cluster level feature gate, a server would know if a feature is enabled on all members, and decide whether or not to accept the new write request.
+
+#### Story 2
+
+A Kubernetes developer would like use a new great etcd feature, but it is not available on all supported etcd versions. What can they do?
+Historically, Kubernetes avoided using any features that were not available on all supported versions of etcd.
+From K8s 1.31, Kubernetes added very complicated and fragile system: [`FeatureSupportChecker`](https://github.com/kubernetes/kubernetes/blob/db82fd1604ebf327ab74cde0a7158a8d95d46202/staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature/feature_support_checker.go#L42) to detect etcd version, parse it and guess whether this feature based on a set of hardcoded etcd version. It does not really know if the feature is enabled by the feature `--experimental` flag.
+
+Now the developer could just send a client request to etcd cluster to query if a feature is enabled on the whole cluster. 
+
+#### Story 3
+
+In a HA cluster, users would like to turn off an enabled feature. They need to restart each server node one by one with feature changed from enabled to disabled. Today, after the restarting process begins and before all nodes are restarted, the behavior of the feature is really undefined: it is enabled on some nodes and disabled on the other nodes.
+
+Now with the cluster feature gate, we can deterministically tell if the feature is enabled or disabled for the whole cluster at any moment.
+
+This story applies similar to the upgrade/downgrade scenarios as well.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Could it done on the client side instead of inside etcd? 
+One could argue that it is easy to know if a feature enabled for the whole cluster by querying each member if the feature is enabled on the client side. But there are several caveates with this approach:
+* not all members are available all the time. if some member is not available, it is questionable if the feature can be used, and that could break the HA assumption. 
+* some feature might change how a raft message is sent and applied, and the order of messages relative to the index the feature is enabled is critical. Here the consistent index is more important than the time, and it is hard to get that outside etcd. 
+
+#### Should we prevent a new member with an inconsistent feature gate from joining the cluster?
+
+One way to prevent inconsistencies of feature gate between different members is to prevent a new member with an inconsistent feature gate from joining the cluster in the first place. 
+We do not think this is the right way because we want to keep the ability to turn on/off a feature in a rolling sequence without bringing down the whole cluster. 
```

Comment:
Would there be a use case where a specific feature requires all the nodes to enable/disable it at the same time in order to function correctly? (thus we have to block inconsistent nodes from joining)

Reply:
There are cases like that. That's why we need the consensus algorithm to set the final values of feature gates. It should not block inconsistent nodes from joining, the leader will decide what value to use.